### ISSUE TITLE
govet can produce false-positive without running go install first

### ISSUE THREAD

**Issue by shuLhan** [golangci/golangci-lint#17]
## Environment

* Go v1.10.2
* Golangci-lint v1.2.1

## Reproduction Steps

* Create Go file with the following content,

```
package main

import "fmt"

func main() {
	format := fmt.Sprintf("%%-%ds %%s", 20)
	fmt.Println("format:", format)
	fmt.Printf(format+"\n", "golangci-lint", "test")
}
```

* Run `golangci-lint run`

## Expected Result

`Congrats! No issues were found.`

## Actual Result

```
master ms 0 % golangci-lint run ./...
Run this tool in cloud on every github pull request in https://golangci.com for free (public repos)
govet.go:8: Printf call has arguments but no formatting directives (govet)
        fmt.Printf(format+"\n", "golangci-lint", "test")
```

**Comment by golangci** [comment]
@shuLhan thank you, it's an interesting case. The most interesting part is:
```
vet: import failed: can't find import: "fmt"
vet: import failed: can't find import: "fmt"
govet.go:3:8: could not import fmt (can't find import: "fmt")
```

Does this program compile?

**Comment by shuLhan** [comment]
> Does this program compile?

Yes.
```
19:36 ~/tmp/sandbox/go/src/bug/govet
master ms 1 % go run govet.go
format: %-20s %s
golangci-lint        test
```
When using `go vet`, it print the following output

```
19:36 ~/tmp/sandbox/go/src/bug/govet
master ms 0 % go vet -v .
bug/govet
# bug/govet
Checking file ./govet.go
./govet.go:8: can't check non-constant format in call to Printf
```
Seems like go vet did not check format if it's not constant.

**Comment by golangci** [comment]
send a full output of next commands, please:
```bash
pwd
echo $GOPATH $GOROOT
cat govet.go
go tool vet -v govet.go
go vet -v .
go get -u github.com/golangci/golangci-lint/cmd/golangci-lint
golangci-lint run -v -E typecheck govet.go
```

**Comment by shuLhan** [comment]
In previous comment I renamed the folder from `tmp/sandbox/go/src/bug/golangci-lint` to `tmp/sandox/go/src/bug/govet`, sorry if it's become confusion, my bad.

Here is the output of all commands,

```
16:58 ~/tmp/sandbox/go/src/bug/govet
master ms 1 % pwd
/home/ms/tmp/sandbox/go/src/bug/govet

16:58 ~/tmp/sandbox/go/src/bug/govet
master ms 0 % echo $GOPATH $GOROOT
/home/ms/tmp/sandbox/go /home/ms/Programs/go

16:59 ~/tmp/sandbox/go/src/bug/govet
master ms 0 % cat govet.go
package main

import "fmt"

func main() {
        format := fmt.Sprintf("%%-%ds %%s", 20)
        fmt.Println("format:", format)
        fmt.Printf(format+"\n", "golangci-lint", "test")
}

16:59 ~/tmp/sandbox/go/src/bug/govet
master ms 0 % go tool vet -v govet.go
vet: import failed: can't find import: "fmt"
vet: import failed: can't find import: "fmt"
govet.go:3:8: could not import fmt (can't find import: "fmt")
Checking file govet.go
govet.go:8: Printf call has arguments but no formatting directives

16:59 ~/tmp/sandbox/go/src/bug/govet
master ms 1 % go vet -v .
bug/govet
# bug/govet
Checking file ./govet.go
./govet.go:8: can't check non-constant format in call to Printf

16:59 ~/tmp/sandbox/go/src/bug/govet
master ms 0 % go run govet.go
format: %-20s %s
golangci-lint        test

16:59 ~/tmp/sandbox/go/src/bug/govet
master ms 0 % golangci-lint run -v govet.go
INFO[0000] Concurrency: 8, machine cpus count: 8
INFO[0000] Active linters: [typecheck structcheck gas varcheck ineffassign deadcode govet errcheck megacheck]
INFO[0000] Found paths for analysis for 9.695µs: [govet.go]
INFO[0000] Program loading took 814.977671ms
INFO[0001] SSA repr building took 164.374438ms
INFO[0001] worker.7 took 4.113µs
INFO[0001] worker.8 took 119.002µs with stages: errcheck: 64.702µs, structcheck: 29.517µs, varcheck: 14.937µs
INFO[0001] worker.6 took 69.045µs with stages: deadcode: 61.899µs
INFO[0001] worker.1 took 126.425µs with stages: typecheck: 119.426µs
INFO[0001] worker.3 took 205.652µs with stages: ineffassign: 199.485µs
INFO[0001] worker.5 took 254.667µs with stages: gas: 246.338µs
INFO[0001] worker.4 took 563.483µs with stages: govet: 544.584µs
govet.go:8:2: Printf call has arguments but no formatting directives (govet)
        fmt.Printf(format+"\n", "golangci-lint", "test")
        ^
INFO[0001] worker.2 took 565.063961ms with stages: megacheck: 565.05436ms
INFO[0001] Workers idle times: #1: 564.854513ms, #3: 564.849004ms, #4: 564.518057ms, #5: 564.807601ms, #6: 564.906788ms, #7: 564.923226ms, #8: 564.912271ms
INFO[0001] processing took 218.749µs with stages: nolint: 106.349µs, exclude: 77.304µs, path_prettifier: 14.445µs, uniq_by_line: 9.889µs, max_same_issues: 4.156µs, max_per_file_from_linter: 2.321µs, max_from_linter: 1.935µs, cgo: 1.494µs, diff: 856ns
INFO[0001] Found 1 issues
INFO[0001] Extracting issued lines took 16.468µs
INFO[0001] Memory: 16 samples, avg is 131.2MB, max is 213.8MB
INFO[0001] Execution took 1.544861173s
```

**Comment by shuLhan** [comment]
I think the `go tool vet` that can't find "fmt" is a bug in Go v1.10.2. I try to build Go from master branch, and there is no error when running `go tool vet`.

```
17:48 ~/tmp/sandbox/go/src/bug/govet
master ms 0 % go version
go version devel +5c36fdfd39 Tue May 29 03:16:17 2018 +0000 linux/amd64

17:49 ~/tmp/sandbox/go/src/bug/govet
master ms 0 % go tool vet -v govet.go
Checking file govet.go
govet.go:8: can't check non-constant format in call to Printf

17:49 ~/tmp/sandbox/go/src/bug/govet
master ms 0 % go vet -v govet.go
command-line-arguments
# command-line-arguments
Checking file ./govet.go
./govet.go:8: can't check non-constant format in call to Printf
```
Recompiling golangci-lint v1.3.4 with Go master, and running it again,
```
17:52 ~/tmp/sandbox/go/src/bug/govet
master ms 0 % golangci-lint run -v govet.go
INFO[0000] Concurrency: 8, machine cpus count: 8
INFO[0000] Active linters: [govet errcheck varcheck deadcode megacheck gas structcheck ineffassign typecheck]
INFO[0000] Found paths for analysis for 11.267µs: [govet.go]
INFO[0000] Program loading took 554.761577ms
INFO[0000] SSA repr building took 181.150492ms
INFO[0000] worker.1 took 102.806µs with stages: deadcode: 98.988µs
INFO[0000] worker.6 took 97.545µs with stages: structcheck: 88.409µs
INFO[0000] worker.2 took 122.522µs with stages: varcheck: 109.051µs, typecheck: 1.03µs
INFO[0000] worker.5 took 268.931µs with stages: ineffassign: 249.447µs
INFO[0000] worker.8 took 342.773µs with stages: gas: 339.58µs
INFO[0000] worker.3 took 689.748µs with stages: errcheck: 672.596µs
INFO[0000] worker.7 took 1.665956ms with stages: govet: 1.653788ms
INFO[0001] worker.4 took 565.249536ms with stages: megacheck: 565.234361ms
INFO[0001] Workers idle times: #1: 565.041153ms, #2: 565.024256ms, #3: 564.46298ms, #5: 564.958687ms, #6: 565.033031ms, #7: 563.581594ms, #8: 564.896172ms
INFO[0001] processing took 13.396µs with stages: max_same_issues: 11.023µs, path_prettifier: 581ns, max_from_linter: 333ns, nolint: 315ns, cgo: 258ns, max_per_file_from_linter: 243ns, diff: 240ns, exclude: 240ns, uniq_by_line: 163ns
Congrats! No issues were found.
INFO[0001] Extracting issued lines took 0s
INFO[0001] Memory: 14 samples, avg is 175.3MB, max is 271.4MB
INFO[0001] Execution took 1.301623595s
```
I think we can conclude that this is issue in Go. Should I close this?

**Comment by golangci** [comment]
@shuLhan I reproduced it by doing cross-compilation in another GOROOT. Fixed it in https://github.com/golangci/golangci-lint/pull/35
Check please in master

**Comment by shuLhan** [comment]
This time I test it in clean GOPATH, no other packages except golangci-lint and test file. I am afraid this is something about cache. Using Go v1.10.2 and golangci-lint v1.3.5.
```
19:39 ~/tmp/sandbox/clean/src/bug/govet
master ms 1 % go version
go version go1.10.2 linux/amd64

19:39 ~/tmp/sandbox/clean/src/bug/govet
master ms 0 % echo $GOPATH $GOROOT
/home/ms/tmp/sandbox/clean /home/ms/Programs/go

19:39 ~/tmp/sandbox/clean/src/bug/govet
master ms 0 % tree -d -L 3 $GOPATH/src
/home/ms/tmp/sandbox/clean/src
├── bug
│   └── govet
└── github.com
    └── golangci
        └── golangci-lint

5 directories

19:39 ~/tmp/sandbox/clean/src/bug/govet
master ms 0 % go vet -v govet.go
command-line-arguments
# command-line-arguments
Checking file ./govet.go
./govet.go:8: can't check non-constant format in call to Printf

19:39 ~/tmp/sandbox/clean/src/bug/govet
master ms 0 % go tool vet -v govet.go
vet: import failed: can't find import: "fmt"
vet: import failed: can't find import: "fmt"
govet.go:3:8: could not import fmt (can't find import: "fmt")
Checking file govet.go
govet.go:8: Printf call has arguments but no formatting directives

19:39 ~/tmp/sandbox/clean/src/bug/govet
master ms 1 % golangci-lint run -v govet.go
INFO[0000] Concurrency: 8, machine cpus count: 8
INFO[0000] Active linters: [gas structcheck ineffassign deadcode govet errcheck megacheck varcheck typecheck]
INFO[0000] set GOROOT="/home/ms/Programs/go"
INFO[0000] Found paths for analysis for 10.582µs: [govet.go]
INFO[0000] Program loading took 638.470008ms
INFO[0000] SSA repr building took 176.29165ms
INFO[0000] worker.3 took 59.8µs with stages: deadcode: 50.032µs, typecheck: 2.133µs
INFO[0000] worker.6 took 51.204µs with stages: varcheck: 42.818µs
INFO[0000] worker.7 took 1.68µs
INFO[0000] worker.1 took 113.405µs with stages: errcheck: 72.739µs, structcheck: 32.897µs
INFO[0000] worker.8 took 176.366µs with stages: ineffassign: 170.01µs
INFO[0000] worker.5 took 253.149µs with stages: gas: 246.608µs
INFO[0000] worker.2 took 440.295µs with stages: govet: 431.26µs
govet.go:8:2: Printf call has arguments but no formatting directives (govet)
        fmt.Printf(format+"\n", "golangci-lint", "test")
        ^
INFO[0001] worker.4 took 554.944407ms with stages: megacheck: 554.932398ms
INFO[0001] Workers idle times: #1: 554.802843ms, #2: 554.513992ms, #3: 554.83425ms, #5: 554.695212ms, #6: 554.814664ms, #7: 554.80942ms, #8: 554.758914ms
INFO[0001] processing took 102.671µs with stages: nolint: 50.749µs, exclude: 35.004µs, path_prettifier: 6.807µs, max_same_issues: 3.29µs, uniq_by_line: 2.706µs, max_from_linter: 1.685µs, max_per_file_from_linter: 1.241µs, cgo: 785ns, diff: 404ns
INFO[0001] Found 1 issues
INFO[0001] Extracting issued lines took 11.463µs
INFO[0001] Memory: 14 samples, avg is 133.2MB, max is 224.5MB
INFO[0001] Execution took 1.370202727s
```

**Comment by golangci** [comment]
1. What if you `export GOROOT="/home/ms/Programs/go"` in shell? Does it work after that?
2. Run `golangci-lint run -v -E typecheck govet.go` please.
3. Maybe it's a bug in `go` or `go vet`, I will dig it.

**Comment by golangci** [comment]
Do you use `gccgo`?

**Comment by golangci** [comment]
@shuLhan try please branch `support/debug-govet-print-false-positive-v2`

**Comment by shuLhan** [comment]
> @shuLhan try please branch support/debug-govet-print-false-positive-v2

Ok, that branch fixed it.

```
0:42 ~
master ms 0 % cd /home/ms/tmp/sandbox/clean/src/github.com/golangci/golangci-lint/

0:42 ~/tmp/sandbox/clean/src/github.com/golangci/golangci-lint
(v1.3.5) ms 0 % git checkout support/debug-govet-print-false-positive-v2
Previous HEAD position was 2f1cdcc Merge pull request #35 from golangci/support/fix-cross-compilation-goroot
Switched to branch 'support/debug-govet-print-false-positive-v2'
Your branch is up to date with 'origin/support/debug-govet-print-false-positive-v2'.

0:43 ~/tmp/sandbox/clean/src/github.com/golangci/golangci-lint
support/debug-govet-print-false-positive-v2 ms 0 % go install ./...

0:43 ~/tmp/sandbox/clean/src/github.com/golangci/golangci-lint
support/debug-govet-print-false-positive-v2 ms 0 % cd $GOPATH/src/bug/govet

0:43 ~/tmp/sandbox/clean/src/bug/govet
master ms 0 % echo $GOROOT $GOPATH $PATH
/home/ms/Programs/go /home/ms/tmp/sandbox/clean /home/ms/tmp/sandbox/clean/bin:/home/ms/.nvm/versions/node/v8.11.1/bin:/home/ms/tmp/sandbox/go/bin:/home/ms/Programs/go_appengine:/home/ms/.python/bin
:/home/ms/.gem/ruby/2.5.0/bin:/home/ms/Programs/android-sdk-linux/platform-tools:/home/ms/Programs/android-sdk-linux/tools:/home/ms/local/bin:/usr/local/bin:/usr/bin:/bin:/usr/local/sbin:/usr/lib/jv
m/default/bin:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl:/home/ms/Programs/go/bin:/home/ms/bin

0:43 ~/tmp/sandbox/clean/src/bug/govet
master ms 0 % which go
/home/ms/Programs/go/bin/go

0:43 ~/tmp/sandbox/clean/src/bug/govet
master ms 0 % which golangci-lint
/home/ms/tmp/sandbox/clean/bin/golangci-lint

0:43 ~/tmp/sandbox/clean/src/bug/govet
master ms 0 % golangci-lint run -v govet.go
INFO[0000] Concurrency: 8, machine cpus count: 8
INFO[0000] Active linters: [structcheck deadcode govet gas errcheck varcheck ineffassign typecheck megacheck]
INFO[0000] set GOROOT="/home/ms/Programs/go"
INFO[0000] Found paths for analysis for 23.9µs: [govet.go]
INFO[0000] Program loading took 606.851083ms
INFO[0000] SSA repr building took 163.691472ms
INFO[0000] worker.2 took 71.992µs with stages: deadcode: 55.613µs, structcheck: 6.108µs, typecheck: 695ns
INFO[0000] worker.1 took 49.28µs with stages: varcheck: 41.896µs
INFO[0000] worker.3 took 98.031µs with stages: errcheck: 90.844µs
INFO[0000] worker.4 took 818ns
INFO[0000] worker.6 took 170.445µs with stages: ineffassign: 161.58µs
INFO[0000] worker.7 took 305.181µs with stages: gas: 292.623µs
INFO[0001] worker.5 took 248.849403ms with stages: govet: 248.831431ms
INFO[0001] worker.8 took 593.109114ms with stages: megacheck: 593.104411ms
INFO[0001] Workers idle times: #1: 592.917923ms, #2: 592.925225ms, #3: 592.908505ms, #4: 592.901597ms, #5: 344.260435ms, #6: 592.895933ms, #7: 592.808421ms
INFO[0001] processing took 8.359µs with stages: max_same_issues: 7.412µs, max_from_linter: 269ns, path_prettifier: 180ns, exclude: 94ns, nolint: 83ns, diff: 81ns, cgo: 81ns, uniq_by_line: 80ns, max_per_file_from_linter: 79ns
Congrats! No issues were found.
INFO[0001] Extracting issued lines took 0s
INFO[0001] Memory: 14 samples, avg is 144.5MB, max is 249.1MB
INFO[0001] Execution took 1.364153386s

0:43 ~/tmp/sandbox/clean/src/bug/govet
master ms 0 % go tool vet -v govet.go
vet: import failed: can't find import: "fmt"
vet: import failed: can't find import: "fmt"
govet.go:3:8: could not import fmt (can't find import: "fmt")
Checking file govet.go
govet.go:8: Printf call has arguments but no formatting directives
```

**Comment by golangci** [comment]
@shuLhan thank you very much! The problem is in default import mechanism of `go vet`: it differs from loading mechanism for another linters and doesn't work for you.
Current fix slows down analysis, it can take a few days to make the final fix.

**Comment by shuLhan** [comment]
@golangci No, thank *you*, for this open source project.

**Comment by jirfag** [comment]
I have an idea as a fast fix for you: try `go install -i ./...` (or `go install -i ./cmd/...`: depends on the project). Does it help?

**Comment by jirfag** [comment]
I suppose it's the same as https://github.com/golangci/golangci-lint/issues/87#issuecomment-396696059
